### PR TITLE
Add action idempotency

### PR DESCRIPTION
**Pull request by danpalmer**

### CODE REVIEW

File: routemaster/state_machine/tests/test_actions.py [hunk 1]
```diff
@@ -62,6 +66,62 @@ def test_actions_do_not_advance_state_on_fail(app_config, create_label, mock_web
     ])
 
 
+def test_actions_retries_use_same_idempotency_token(app_config, create_label, mock_webhook, assert_history):
```
Comment (PeterJCLaw) [inline routemaster/state_machine/tests/test_actions.py]: Would be good to have a test which validates the inverse -- that different actions get different tokens.

File: routemaster/state_machine/tests/test_actions.py [hunk 2]
```diff
@@ -62,6 +66,62 @@ def test_actions_do_not_advance_state_on_fail(app_config, create_label, mock_web
...
+    state_machine = app_config.config.state_machines['test_machine']
+
+    # Figure out what our idempotency token should be
```
Comment (PeterJCLaw) [inline routemaster/state_machine/tests/test_actions.py]: this couples the test to the implementation; I'd rather have a test which learns what it should be from the first action and then ensures that the second call matches the first.

Comment (danpalmer) [reply routemaster/state_machine/tests/test_actions.py]: Yeah I wanted to do that but it looked quite tricky. I'll see what I can do.

Comment (danpalmer) [reply routemaster/state_machine/tests/test_actions.py]: fe9dc27608b651d92912681589bf12f6744a6668

File: routemaster/state_machine/tests/test_actions.py (hunk 1 above)
Comment (danpalmer) [reply routemaster/state_machine/tests/test_actions.py]: Added in fe9dc27608b651d92912681589bf12f6744a6668

File: routemaster/state_machine/tests/test_actions.py (hunk 2 above)
Comment (PeterJCLaw) [reply routemaster/state_machine/tests/test_actions.py]: Ah, in the past I've used [`call_args`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.call_args) or [`call_args_list`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.call_args_list) for getting the arguments from the first call. Unfortunately that ends up being about as complex as you need to cope with them being empty in the failure cases.

File: routemaster/state_machine/tests/test_actions.py [hunk 3]
```diff
@@ -62,6 +66,104 @@ def test_actions_do_not_advance_state_on_fail(app_config, create_label, mock_web
...
+    state_machine = app_config.config.state_machines['test_machine']
+
+    seen_tokens = []
+
+    def persist_token(url, content_type, data, token):
+        seen_tokens.append(token)
+        return WebhookResult.SUCCESS
+
+    # First get the label into the action state by failing the automatic
+    # progression through the machine.
```
Comment (PeterJCLaw) [inline routemaster/state_machine/tests/test_actions.py]: I'm not sure this comment is accurate -- `persist_token` returns `WebhookResult.SUCCESS` -- is that expected?

File: routemaster/state_machine/tests/test_actions.py [hunk 4]
```diff
@@ -62,6 +66,104 @@ def test_actions_do_not_advance_state_on_fail(app_config, create_label, mock_web
...
+
+    webhook.assert_called_once_with(
+        'http://localhost/hook',
+        'application/json',
+        b'{"label": "foo", "metadata": {"should_progress": true}}',
+        expected['token'],
+    )
+
+
+def test_different_actions_use_different_idempotency_tokens(app_config, create_label, mock_webhook, assert_history):
```
Comment (PeterJCLaw) [inline routemaster/state_machine/tests/test_actions.py]: I don't seen an assertion here that the tokens are different, only that there are three entries in the `seen_tokens` _list_.

File: routemaster/tests/test_webhook_runner.py [hunk 5]
```diff
@@ -118,7 +118,22 @@ def test_requests_webhook_runner_for_state_machine_does_not_apply_headers_for_no
...
     runner = webhook_runner_for_state_machine(state_machine)
-    runner('http://not-example.com', 'application/test-data', b'\0\xff')
+    runner('http://not-example.com', 'application/test-data', b'\0\xff', '')
 
     last_request = httpretty.last_request()
     assert last_request.headers['x-api-key'] is None
+
+
+@httpretty.activate
+def test_requests_webhook_runner_passes_idempotency_token_through():
```
Comment (PeterJCLaw) [inline routemaster/tests/test_webhook_runner.py]: :+1:

File: routemaster/state_machine/tests/test_actions.py (hunk 2 above)
Comment (PeterJCLaw) [reply routemaster/state_machine/tests/test_actions.py]: for clarity: no need to change this

File: routemaster/state_machine/actions.py [hunk 6]
```diff
@@ -88,6 +94,12 @@ def _process_action_with_metadata(
     return True
 
 
+def _calculate_idempotency_token(label: LabelRef, history_id: int):
+    return hashlib.sha256(
+        f"{label.name}:{history_id}".encode('ascii'),
```
Comment (prophile) [inline routemaster/state_machine/actions.py]: I don't believe `name` is restricted to being ASCII elsewhere; would advocate for encoding this in UTF-8.

File: routemaster/state_machine/actions.py [hunk 7]
```diff
@@ -88,6 +94,12 @@ def _process_action_with_metadata(
     return True
 
 
+def _calculate_idempotency_token(label: LabelRef, history_id: int):
```
Comment (prophile) [inline routemaster/state_machine/actions.py]: Could annotate the return type here.

File: routemaster/webhooks.py [hunk 8]
```diff
@@ -16,7 +16,7 @@ class WebhookResult(enum.Enum):
     FAIL = 'fail'
 
 
-WebhookRunner = Callable[[str, str, bytes], WebhookResult]
+WebhookRunner = Callable[[str, str, bytes, str], WebhookResult]
```
Comment (prophile) [inline routemaster/webhooks.py]: We should probably `NewType` these. With four arguments this is beginning to get unwieldy.

File: routemaster/state_machine/tests/test_actions.py [hunk 9]
```diff
@@ -62,6 +66,104 @@ def test_actions_do_not_advance_state_on_fail(app_config, create_label, mock_web
...
+
+    # First get the label into the action state by failing the automatic
+    # progression through the machine.
+    with mock.patch(
+        'routemaster.webhooks.RequestsWebhookRunner.__call__',
+        side_effect=persist_token,
+    ) as webhook:
+        create_label('foo', state_machine.name, {'should_progress': True})
+        create_label('bar', state_machine.name, {'should_progress': True})
+        create_label('baz', state_machine.name, {'should_progress': True})
```
Comment (prophile) [inline routemaster/state_machine/tests/test_actions.py]: Seems a shame we have to resort to mocking when the webhook runner is something that is passed in. Could we attach it to the app config?

File: routemaster/state_machine/actions.py (hunk 7 above)
Comment (prophile) [reply routemaster/state_machine/actions.py]: A docstring would also be nice (explaining a little of the role of idempotency tokens and their semantics).

File: routemaster/state_machine/tests/test_actions.py (hunk 4 above)
Comment (danpalmer) [reply routemaster/state_machine/tests/test_actions.py]: Whoops!! f8659f73079ff5fb150883cc7481b50111e1b43b

File: routemaster/state_machine/tests/test_actions.py (hunk 3 above)
Comment (danpalmer) [reply routemaster/state_machine/tests/test_actions.py]: 8d81ee7675be0c70103688d9684f2359ca8d07c9

File: routemaster/state_machine/actions.py (hunk 7 above)
Comment (danpalmer) [reply routemaster/state_machine/actions.py]: I've generally tidied this up a lot in 3e7b914b5b2078ccc710e251a560b59b1c3cee15

File: routemaster/state_machine/actions.py (hunk 6 above)
Comment (danpalmer) [reply routemaster/state_machine/actions.py]: Yeah I just dropped the label name. History IDs are unique. 3e7b914b5b2078ccc710e251a560b59b1c3cee15

File: routemaster/webhooks.py (hunk 8 above)
Comment (danpalmer) [reply routemaster/webhooks.py]: Unfortunately it doesn't look like you can use `NewType` like this, or I'm reading the docs wrong and using it incorrectly.

File: routemaster/state_machine/tests/test_actions.py (hunk 9 above)
Comment (danpalmer) [reply routemaster/state_machine/tests/test_actions.py]: This needs more thought as I've now messed up the API with the webhook configs.

I'm going to:

 - Pull the constructor out of the `RequestsWebhookRunner`, making `__call__` instead take an optional dict of headers.
 - Pull the header selection logic out onto the config model objects in a `headers_for_url` method.
 - Add the webhook runner to the app config class.
 - Add some methods to the test app config to control the behaviour and remove the mocking.

Also I want to extend `create_label` in tests to take an optional state, and create the history, but not actually execute the states for the path to that state. This should simplify the tests considerably.

I'll do all of this in a separate PR.

Comment (danpalmer) [reply routemaster/state_machine/tests/test_actions.py]: #23

**Pull request merged**